src/Raven.Client/Util/CertificateLoaderUtil.cs

```diff
@@ -0,0 +1,133 @@
+        }
+        catch (Exception e)
+        {
+            exception = e;
+            f = SwitchUserKeySetWithMachineKeySet(f);
+            collection.Import(rawData, password, f);
+        }
+
+        LogIfNeeded(nameof(Import), f, exception);
+    }
+
+    public static X509Certificate2 CreateCertificate(byte[] rawData, string password = null, X509KeyStorageFlags? flags = null)
+    {
+        return CreateCertificate(f => new X509Certificate2(rawData, password, f), flags);
+    }
+
+    internal static X509Certificate2 CreateCertificate(string fileName, string password = null, X509KeyStorageFlags? flags = null)
+    {
+        return CreateCertificate(f => new X509Certificate2(fileName, password, f), flags);
+    }
+
+    private static X509Certificate2 CreateCertificate(Func<X509KeyStorageFlags, X509Certificate2> creator, X509KeyStorageFlags? flag)
+    {
+        DebugAssertDoesntContainKeySet(flag);
+        var f = AddUserKeySet(flag);
+        Exception exception = null;
+        X509Certificate2 certificate;
+
+        try
+        {
+            certificate = creator(f);
+        }
+        catch(Exception e)
+        {
+            exception = e;
+            f = SwitchUserKeySetWithMachineKeySet(f);
+            certificate = creator(f);
+        }
+        
+        LogIfNeeded(nameof(CreateCertificate), f, exception);
+
+        CertificateCleaner.RegisterForDisposalDuringFinalization(certificate);
+
+        return certificate;
+    }
+
+    private static X509KeyStorageFlags SwitchUserKeySetWithMachineKeySet(X509KeyStorageFlags f)
+    {
+        return f & ~X509KeyStorageFlags.UserKeySet | X509KeyStorageFlags.MachineKeySet;
+    }
+
+    private static X509KeyStorageFlags AddUserKeySet(X509KeyStorageFlags? flag)
+    {
+        return (flag ?? X509KeyStorageFlags.DefaultKeySet) | X509KeyStorageFlags.UserKeySet;
+    }
+
+    [Conditional("DEBUG")]
+    private static void DebugAssertDoesntContainKeySet(X509KeyStorageFlags? flags)
+    {
+        const X509KeyStorageFlags keyStorageFlags =
+#if NETCOREAPP3_1_OR_GREATER 
+            X509KeyStorageFlags.EphemeralKeySet |
+#endif
+            X509KeyStorageFlags.UserKeySet |
+            X509KeyStorageFlags.MachineKeySet;
+        
+        Debug.Assert(flags.HasValue == false || (flags.Value & keyStorageFlags) == 0);
+    }
+    
+    private static void LogIfNeeded(string method, X509KeyStorageFlags f, Exception exception)
+    {
+        if (FirstTime)
+        {
+            FirstTime = false;
+            if (Log.IsOperationsEnabled)
+                Log.Operations(CreateMsg());
+        }
+        else
+        {
+            if (Log.IsInfoEnabled)
+                Log.Info(CreateMsg());
+        }
+
+        string CreateMsg()
+        {
+            var msg = $"{nameof(CertificateLoaderUtil)}.{method} - Flags used {f}";
+            if (exception != null)
+                msg += $"First attempt exception : {exception}";
+            return msg;
+        }
```

Comment:
We have support for outputting the exception message _separately_ for a reason.
For example, it indicates an `error` when using debug log viewer in the studio, let's use the API properly